crates/kv-store/src/sstable.rs

```diff
@@ -0,0 +1,1026 @@
+                (idx, iter, Some(start))
+            }
+            Bound::Unbounded => {
+                let block = table.read_block_cached(0);
+                let iter = BlockIter::new_seek_to_first(block);
+                (0, iter, None)
+            }
+        };
+        let (end_idx, end_iter, end_excluded) = match end {
+            Bound::Included(end) => {
+                let end_idx = table.find_back_block_idx(end);
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), None)
+                }
+            }
+            Bound::Excluded(end) => {
+                let end_idx = table.find_back_block_idx(end);
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, Some(end))
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), Some(end))
+                }
+            }
+            Bound::Unbounded => {
+                let end_idx = table.meta.len() - 1;
+                if end_idx == table_idx {
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_seek_to_first(block);
+                    (end_idx, Some(iter), None)
+                }
+            }
+        };
+
+        let mut ans = if let Some(end_iter) = end_iter {
+            debug_assert!(end_idx > table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter,
+                next_block_idx: table_idx,
+                back_block_iter: end_iter,
+                back_block_idx: end_idx as isize,
+                next_first: false,
+            }
+        } else {
+            debug_assert!(end_idx == table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter.clone(),
+                next_block_idx: table_idx,
+                back_block_iter: iter,
+                back_block_idx: end_idx as isize,
+                next_first: true,
+            }
+        };
+        // the current iter is empty, but has next iter. we need to skip the empty iter
```

Comment:
Why current iter is empty?

Reply:
The comment is wrong, It should be `the current iter may be empty`